### ISSUE TITLE
V0.1.39 rapidly increases RAM usage till it crashes

### ISSUE THREAD

**Issue by crProductGuy** [iron-fish/ironfish#1848]
### What happened?

Upgraded to 0.1.39 via npm on win10 on my cheap laptop with 2-core Athlon silver, running a dedicated node (no mining).  Total usable RAM is 5.9GB.   Iron Fish starts about 2.4 GB, and expands until RAM full and JS crash.
[4x JS mark-sense compact OOM errors on 0.1.39 in first 12 hours of running it..txt](https://github.com/iron-fish/ironfish/files/9172510/4x.JS.mark-sense.compact.OOM.errors.on.0.1.39.in.first.12.hours.of.running.it.txt)

Cannot run node longer than 30-45 min without JavaScript out of memory error.

Examples in attached file.

Last crash:
Found peer pPnsDlP (DaBusy) ancestor 00000...ff7b2, syncing from 104039 -> 105612 (1573) after 1 requests
Requesting 10 blocks starting at 00000...ff7b2 (104039) from pPnsDlP (DaBusy)
Added block seq: 104040, hash: 00000...e4034, txs: 6, progress: 98.49%, time: 4501.3ms

<--- Last few GCs --->

[12796:0000017C9D1C6900]  1412435 ms: Mark-sweep 1975.3 (2085.1) -> 1960.5 (2085.9) MB, 3468.7 / 1.8 ms  (average mu = 0.159, current mu = 0.079) allocation failure scavenge might not succeed
[12796:0000017C9D1C6900]  1416250 ms: Mark-sweep 1979.8 (2089.2) -> 1964.1 (2089.4) MB, 3499.6 / 2.4 ms  (average mu = 0.122, current mu = 0.083) allocation failure scavenge might not succeed

<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 00007FF6005A79CF v8::internal::CodeObjectRegistry::~CodeObjectRegistry+114207
 2: 00007FF600536096 DSA_meth_get_flags+65542
 3: 00007FF600536F4D node::OnFatalError+301
 4: 00007FF600E6B1EE v8::Isolate::ReportExternalAllocationLimitReached+94
 5: 00007FF600E557CD v8::SharedArrayBuffer::Externalize+781
 6: 00007FF600CF8B9C v8::internal::Heap::EphemeronKeyWriteBarrierFromCode+1468
 7: 00007FF600D05849 v8::internal::Heap::PublishPendingAllocations+1129
 8: 00007FF600D0281A v8::internal::Heap::PageFlagsAreConsistent+2842
 9: 00007FF600CF5479 v8::internal::Heap::CollectGarbage+2137
10: 00007FF600CF3630 v8::internal::Heap::AllocateExternalBackingStore+2000
11: 00007FF600D181B6 v8::internal::Factory::NewFillerObject+214
12: 00007FF600A4A685 v8::internal::DateCache::Weekday+1797
13: 00007FF600EF8EE1 v8::internal::SetupIsolateDelegate::SetupHeap+494417
14: 0000017C9F14967E

Jason confirmed in #report-bugs chat on 22july that it's an issue with 0.1.39 as confirmed on Iron Fish systems.  

Symptom: JS heap increases without bounds until RAM capacity of host machine is exhausted, then the node process dies.

last loop of 2-min Ironfish status command:
Version              0.1.39 @ 277a1b1
Node                 STARTED
Memory               Heap: 1.83 GiB / 1.92 GiB (95.3%), RSS: 2.58 GiB (43.9%), Free: 602.93 MiB (90.0%)
P2P Network          CONNECTED - In: 933.66 KB/s, Out: 1.38 KB/s, peers 44
Mining               STARTED - 0 miners, 0 mined
Mem Pool             0 tx
Syncer               SYNCING - 0.09 blocks per seconds, avg time to add block 6941.14 ms, progress: 98.49%
Blockchain           NOT SYNCED @ HEAD 000000000000ab5176cb188c8fdf24d09eb2445009165d1cf289622f039e1089 (104038)
Telemetry            STARTED - 229 <- 0 pending
Workers              STARTED - 123 -> 1 / 1 - -0.45 jobs Δ, 29.44 jobs/s
sleeping-120-sec
date+time
Friday, July 22, 2022 3:14:35 PM

Version              0.1.39 @ 277a1b1
Node                 STOPPED
Memory               Heap: 0 B / 0 B (NaN%), RSS: 0 B (0.0%), Free: 0 B (100.0%)

### Version

v0.1.39

### Debug output

```shell
Iron Fish version            0.1.39 @ 277a1b1
Iron Fish library            0.0.17 @ 277a1b1
Operating system             Windows_NT x64
CPU model(s)                 AMD Athlon Silver 3050U with Radeon Graphics
CPU threads                  2
RAM total                    5.88 GiB
Node version                 v16.15.1
ironfish in PATH             true
Telemetry enabled            true
Accounts head hash           0000000000093fa29ef321e0cd6ffdcd87e1a40302331c41dd7212b5ab5ff7b2
Accounts head in chain       true
Accounts head sequence       104039
```

### Relevant log output

```shell
Found peer pPnsDlP (DaBusy) ancestor 00000...ff7b2, syncing from 104039 -> 105612 (1573) after 1 requests
Requesting 10 blocks starting at 00000...ff7b2 (104039) from pPnsDlP (DaBusy)
Added block seq: 104040, hash: 00000...e4034, txs: 6, progress: 98.49%, time: 4501.3ms

<--- Last few GCs --->

[12796:0000017C9D1C6900]  1412435 ms: Mark-sweep 1975.3 (2085.1) -> 1960.5 (2085.9) MB, 3468.7 / 1.8 ms  (average mu = 0.159, current mu = 0.079) allocation failure scavenge might not succeed
[12796:0000017C9D1C6900]  1416250 ms: Mark-sweep 1979.8 (2089.2) -> 1964.1 (2089.4) MB, 3499.6 / 2.4 ms  (average mu = 0.122, current mu = 0.083) allocation failure scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 00007FF6005A79CF v8::internal::CodeObjectRegistry::~CodeObjectRegistry+114207
 2: 00007FF600536096 DSA_meth_get_flags+65542
 3: 00007FF600536F4D node::OnFatalError+301
 4: 00007FF600E6B1EE v8::Isolate::ReportExternalAllocationLimitReached+94
 5: 00007FF600E557CD v8::SharedArrayBuffer::Externalize+781
 6: 00007FF600CF8B9C v8::internal::Heap::EphemeronKeyWriteBarrierFromCode+1468
 7: 00007FF600D05849 v8::internal::Heap::PublishPendingAllocations+1129
 8: 00007FF600D0281A v8::internal::Heap::PageFlagsAreConsistent+2842
 9: 00007FF600CF5479 v8::internal::Heap::CollectGarbage+2137
10: 00007FF600CF3630 v8::internal::Heap::AllocateExternalBackingStore+2000
11: 00007FF600D181B6 v8::internal::Factory::NewFillerObject+214
12: 00007FF600A4A685 v8::internal::DateCache::Weekday+1797
13: 00007FF600EF8EE1 v8::internal::SetupIsolateDelegate::SetupHeap+494417
14: 0000017C9F14967E
```

### Graffiti

ProductGuy

**Comment by moredillon** [comment]
Had the same thing, "JavaScript heap out of memory" repeated crashes happen after updating to 1.39 from 1.37
I am running a node, pool, and mining on this machine. Never had this issue before.
Running on Win 10 (powershell/ubuntu) with 128 GB RAM and this is after fresh restart.

```
<--- Last few GCs --->

[164:0x63ad780]  4104684 ms: Mark-sweep (reduce) 3922.4 (4143.6) -> 3920.5 (4143.3) MB, 6208.6 / 4.7 ms  (average mu = 0.100, current mu = 0.022) allocation failure scavenge might not succeed
[164:0x63ad780]  4109304 ms: Mark-sweep (reduce) 3921.0 (4143.8) -> 3920.7 (4143.8) MB, 4601.3 / 4.7 ms  (average mu = 0.061, current mu = 0.004) allocation failure GC in old space requested


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xb09c10 node::Abort() [/usr/bin/node]
 2: 0xa1c193 node::FatalError(char const*, char const*) [/usr/bin/node]
 3: 0xcf8dbe v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/usr/bin/node]
 4: 0xcf9137 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/bin/node]
 5: 0xeb09d5  [/usr/bin/node]
 6: 0xeb14b6  [/usr/bin/node]
 7: 0xebf9de  [/usr/bin/node]
 8: 0xec0420 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/bin/node]
 9: 0xec339e v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/bin/node]
10: 0xe84612 v8::internal::Factory::AllocateRaw(int, v8::internal::AllocationType, v8::internal::AllocationAlignment) [/usr/bin/node]
11: 0xe7ef2c v8::internal::FactoryBase<v8::internal::Factory>::AllocateRawArray(int, v8::internal::AllocationType) [/usr/bin/node]
12: 0xe7f005 v8::internal::FactoryBase<v8::internal::Factory>::NewFixedArrayWithFiller(v8::internal::Handle<v8::internal::Map>, int, v8::internal::Handle<v8::internal::Oddball>, v8::internal::AllocationType) [/usr/bin/node]
13: 0x10e788e v8::internal::MaybeHandle<v8::internal::OrderedHashMap> v8::internal::OrderedHashTable<v8::internal::OrderedHashMap, 2>::Allocate<v8::internal::Isolate>(v8::internal::Isolate*, int, v8::internal::AllocationType) [/usr/bin/node]
14: 0x10e7943 v8::internal::MaybeHandle<v8::internal::OrderedHashMap> v8::internal::OrderedHashTable<v8::internal::OrderedHashMap, 2>::Rehash<v8::internal::Isolate>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::OrderedHashMap>, int) [/usr/bin/node]
15: 0x11f3a55 v8::internal::Runtime_MapGrow(int, unsigned long*, v8::internal::Isolate*) [/usr/bin/node]
16: 0x15f2099  [/usr/bin/node]
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by yuanaichi** [comment]
I had a same issue

**Comment by danield9tqh** [comment]
This was fixed in https://github.com/iron-fish/ironfish/pull/1851 and released in v0.1.40. Thanks for the report!
